File: cpp Engine/src/engine.cpp
#include "engine.hpp"
#include <random>

Engine::Engine(Board* board) : board(board) {}

Move Engine::get_best_move(int depth) {
    int alpha = -MATE, beta = +MATE;
    int best_score = -MATE;

    auto moves = board->get_moves();
    if (moves.empty()) return Move{};
    
    // Reserve capacity to avoid reallocationsa
    moves.reserve(256);

    std::sort(moves.begin(), moves.end(),
              [this](const Move& a, const Move& b){ return score_move(a) > score_move(b); });

    std::vector<Move> best_moves;
    best_moves.reserve(10); // Reserve space for best moves

    for (auto& move : moves) {
        board->make_move(&move);
        int score = -minimax(depth - 1, -beta, -alpha);
        board->undo_move();
        std::cout << "move: " << move.to_string() << " score: " << score << "\n";

        if (score > best_score) {
            best_score = score;
            best_moves.clear();
            // best_moves.emplace_back(move);
            best_moves.push_back(move);
        } else if (score == best_score) {
            // best_moves.emplace_back(move);
            best_moves.push_back(move);
        }

        if (score > alpha) alpha = score;
    }

    for (auto& move : best_moves) {
        std::cout << "\tmove: " << move.to_string() << "\n";
    }

    if (best_moves.size() == 1) return best_moves[0];

    static thread_local std::mt19937 rng(std::random_device{}());
    static thread_local std::uniform_int_distribution<size_t> dist;
    dist.param(std::uniform_int_distribution<size_t>::param_type(0, best_moves.size() - 1));
    return best_moves[dist(rng)];
}

int Engine::score_move(Move move) {
    int end = move.end();
    Piece move_piece = board->get_piece(move.start());
    Piece end_piece = board->get_piece(end);
    Piece promotion_piece = move.promotion_piece(board->get_turn());
    uint8_t move_piece_type = move_piece.get_piece();
    uint8_t end_piece_type = end_piece.get_piece();
    uint8_t promotion_piece_type = promotion_piece.get_piece();

    int score = 0;
    if (!end_piece.is_empty()) {
        score += 10 * PIECE_VALUES[end_piece_type] - PIECE_VALUES[move_piece_type];
    }
    if (!promotion_piece.is_empty()) {
        score += PIECE_VALUES[promotion_piece_type];
    }
    if (board->is_controlled(end)) {
        score -= PIECE_VALUES[move_piece_type];
    }
    return score;
}

int Engine::search(int depth) {
    if (depth == 0) {
        return 1;
    }
    int ret = 0;
    std::vector<Move> moves = board->get_moves();
    moves.reserve(256); // Reserve capacity for typical chess positions
    for (const Move& move : moves) {
        board->make_move(&move);
        ret += search(depth - 1);
        board->undo_move();
    }
    return ret;
}

int Engine::evaluate() {
    GameState game_state = board->get_game_state();
    if (game_state == GameState::WHITE_WIN || game_state == GameState::BLACK_WIN) {
        return -MATE;
    } else if (game_state == GameState::DRAW) {
        return 0;
    }

    int ret = 0;

    // material values
    for (int i = 0; i < 6; ++i) {
        ret += __builtin_popcountll(board->friend_arr[i]) * PIECE_VALUES[i];
        ret -= __builtin_popcountll(board->enemy_arr[i]) * PIECE_VALUES[i];
    }

    // position values
    auto flip = [](int sq){ return 63 - sq; };
    Turn currentTurn = board->get_turn();
    bool whiteToMove = (currentTurn == Turn::WHITE);

    int sq;
    for (int i = 0; i < 6; ++i) {
        CTZLL_ITERATOR(sq, board->friend_arr[i]) {
            ret += POSITION_VALUES[i][ whiteToMove ? sq : flip(sq) ];
        }
        CTZLL_ITERATOR(sq, board->enemy_arr[i]) {
            ret -= POSITION_VALUES[i][ whiteToMove ? flip(sq) : sq ];
        }
    }

    return ret;
}

int Engine::minimax(int depth, int alpha, int beta) {
    if (depth == 0) {
        return evaluate();
    }

    if (board->get_game_state() != GameState::IN_PROGRESS) {
        return evaluate();
    }

    std::vector<Move> moves = board->get_moves();
    std::sort(moves.begin(), moves.end(), 
        [this](const Move& a, const Move& b) {
        return score_move(a) > score_move(b);
    });

    int score;
    for (const Move& move : moves) {
        board->make_move(&move);
        score = -minimax(depth - 1, -beta, -alpha);
        board->undo_move();

        if (score >= beta) {
            return beta;
        }
        alpha = std::max(alpha, score);
    }
    return alpha;
}

File: Python Engine/engine.py
from utils import *
from board import Board, GameState, Turn
from position import Position

class Engine:
    def __init__(self):
        self.positions = 0
        pass

    def evaluate(self, board: Board) -> int:
        board._ensure_updated()
        if board.game_state == GameState.WHITE or board.game_state == GameState.BLACK:
            return float('-inf')
        elif board.game_state == GameState.DRAW:
            return 0

        ret = 0

        # add and subtract material
        for p, val in zip(board.friend_list, PIECE_VALUES):
            ret += p.bit_count() * val
        for p, val in zip(board.enemy_list, PIECE_VALUES):
            ret -= p.bit_count() * val
        
        # add and subtract based off of position
        # for p, values in zip(board.friend_list, POSITION_VALUES):
        #     for n in ctzll_iterator(p):
        #         ret += values[n] if board.turn == Turn.WHITE else values[63 - n]
        # for p, values in zip(board.enemy_list, POSITION_VALUES):
        #     for n in ctzll_iterator(p):
        #         ret -= values[n] if board.turn == Turn.WHITE else values[63 - n]
        return ret

    def search(self, board: Board, depth: int) -> int:
        if depth == 0:
            return 1
        ret = 0
        for s, e, piece, promo in board.get_all_moves_list():
            taken, state_changes = board.make_move(s, e, piece, promo)
            ret += self.search(board, depth - 1)
            board.undo_move(s, e, piece, taken, state_changes)
        return ret
    
    def score_move(self, board: Board, piece: Piece, e: Position, promo: Optional[Piece]) -> int:
        guess = 0
        if board.board[e.num] != Piece.EMPTY:
            guess += 10 * PIECE_VALUES[board.board[e.num] % COLOR_DELIMITER] - PIECE_VALUES[piece]
        if promo is not None:
            guess += PIECE_VALUES[promo]
        if e.overlap(board.controlled):
            guess -= PIECE_VALUES[piece]
        return guess

    def minimax(self, board: Board, depth: int, alpha: int, beta: int) -> int:
        if depth == 0:
            self.positions += 1
            return self.evaluate(board)
        
        board._ensure_updated()
        if board.game_state == GameState.WHITE or board.game_state == GameState.BLACK:
            self.positions += 1
            return float('-inf')
        elif board.game_state == GameState.DRAW:
            self.positions += 1
            return 0
        
        all_moves = board.get_all_moves_list()
        all_moves.sort(key=lambda x: self.score_move(board, x[2], x[1], x[3]),reverse=True)
        for s, e, piece, promo in all_moves:
            taken, state_changes = board.make_move(s, e, piece, promo)
            curr_eval = -self.minimax(board, depth - 1, -beta, -alpha)
            board.undo_move(s, e, piece, taken, state_changes)

            if curr_eval >= beta:
                return beta
            alpha = max(alpha, curr_eval)
        return alpha

if __name__ == "__main__":
    li = Engine()
    board = Board()

    # for i in range(1, 6):
    #     a = time.time_ns()
    #     li.positions = 0
    #     engine_eval = li.minimax(board, i, float('-inf'), float('inf'))
    #     b = time.time_ns()
    #     print(f"depth={i}: evaluated {li.positions} positions in {(b - a) // 10**6} ms")
    
    for i in range(1, 5):
        a= time.time_ns()
        positions = li.search(board, i)
        b = time.time_ns()
        print(f"depth={i}: calculated {positions} in {(b - a) // 10**6} ms")